op-batcher/batcher/channel_manager.go

```diff
@@ -126,11 +126,21 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
 // in the block queue and the blockCursor is ahead of it.
 // Panics if the block is not in state.
 func (s *channelManager) rewindToBlock(block eth.BlockID) {
+	initialCursor := s.blockCursor
 	idx := block.Number - s.blocks[0].Number().Uint64()
 	if s.blocks[idx].Hash() == block.Hash && idx < uint64(s.blockCursor) {
 		s.blockCursor = int(idx)
 	} else {
-		panic("tried to rewind to nonexistent block")
+		panic("rewindToBlock: tried to rewind to nonexistent block")
+	}
+
+	// Ensure metrics stay in sync by re-adding blocks which the cursor rewound over
+	for i := initialCursor - 1; i >= s.blockCursor; i-- {
+		block, ok := s.blocks.PeekN(i)
+		if !ok {
+			panic("rewindToBlock: block not found at index " + fmt.Sprint(i))
+		}
+		s.metr.RecordL2BlockInPendingQueue(block)
```

Comment:
For future work: instead of the `types.Block`s directly, we could save a wrapper type in the blocks queue that caches the estimated DA size. Because when we requeue 1000s of blocks here, we also have to perform the FastLZ compression on all of them again. So if we can cache them the first time we do it, we would remove some stress on the batcher during requeue events.

Reply:
Yeah good idea, I added a link to it to #14559